Enable p refinement

YaqiWang opened the pull request:
Close  #24141.

YaqiWang committed:
enable p refinement #24141

YaqiWang commented:
@roystgnr and @lindsayad I added the call to `switch_h_to_p_refinement()` but unfortunately, this does not work. The test I tried hit an error
```
We caught a libMesh error in ThreadedElementLoopBase:ERROR: Bad ElemType = QUAD4 for SECOND order approximation!
```
which does not quite make sense to me. Looks like the code somehow assumes continuous FEM somewhere. Any ideas?

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1483058/) on 107b880 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/24180/site)

This comment will be updated on new commits.

roystgnr commented:
I don't see any explicitly declared continuous variables, but at the libMesh level we never just add a variable unless it's requested by a higher level.  Looks like we are in the nonlinear system, too, not postprocessing or something.  Based on the stack trace at the libMesh error, MOOSE seems to have something in `_fe` that it's trying to reinit but shouldn't be.

I can reproduce this; let me dig in.

roystgnr commented:
Ah.  I didn't have to look far.  The very first thing the `Assembly` constructor does is "build fe's for the helpers", with type LAGRANGE, for every dimension, so that we'll reinit every Lagrange shape function on every element whether we're ever going to use it or not.

Someone's definition of "helping" clearly differs from mine.

roystgnr commented:
Ah, and naturally we request phi and dphi and xyz and JxW from all the element helpers, and all that plus normals from all the face helpers.

When we finally figure out how to automate regression testing for performance, I think "throw in d2phi and tangents too for giggles" would be a good test PR to make sure we catch true regressions, and "**only do this when we need it**" would be a good way to make sure we spotlight improvements.

YaqiWang commented:
@roystgnr Thanks for looking into this!

roystgnr commented:
I'd love to accept your thanks, if only I could tell you how to fix it.

Bumping up element order in an input file isn't going to help; that'll just tell `Assembly` to create the helpers as `SECOND` rather than `FIRST` order, and then they'll complain when p refinement wants to bump them up to `THIRD` and *that's* not supported.

Getting rid of these helper FE when they're not needed would be a big performance fix in some cases, but there are probably other cases where specifically-LAGRANGE helper objects *are* needed - DG on a displaced mesh, maybe - so we can't rely on that as a fix here.

In libMesh we'll often hack a p-level down to 0 temporarily to get a "base" reinit, but that's not a good option for `Assembly` - no telling when two different threads will call e.g. `Assembly::reinitNeighbor` in such a way as to try to work on the same element at the same time.

I think the best fix is to get proper "ignore p refinement" flags into all the appropriate libMesh APIs (right now we can operate that way with some but not others, because the ugly hack worked in the handful of places I needed it...), then we can use those for the MOOSE helper FEs.  It'll be at least a week before I can do that and then at least another before we get it downstream, though.

YaqiWang commented:
FYI: Just removed the unnecessary marker I checked in the commit.

lindsayad commented:
> I think the best fix is to get proper "ignore p refinement" flags into all the appropriate libMesh APIs (right now we can operate that way with some but not others, because the ugly hack worked in the handful of places I needed it...

What is this ugly hack? And what are the "some" that I can look to for mimicking?

roystgnr commented:
> What is this ugly hack?

I was at least self aware about it: `git grep hack` soon finds `geom/elem.h:void hack_p_level (const unsigned int p);`  (I should have also put "ugly" in the API, though.)  We even have to **const_cast** to use this thing in some places.  A few use cases are reasonable (we use it inside `set_p_level` for better factoring, and we use it in element copy / communication / IO code to skip `set_p_level` checks when we're in the middle of changing a swath of raw settings), but there's an idiom in HPCoarsenTest, and FEBase even, where we hack a p level temporarily and then hack it back after brief calculations.

> And what are the "some" that I can look to for mimicking?

`git grep add_p_level` ... not sure how to easily find the use cases of that parameter, but at least the behavior of it is clear: in individual calls to shape (and derivative, second derivative, etc) calculations, we can tell FE specializations that they should ignore p elevation for the purposes of that call.  But for some reason I never propagated this up to something like a parameter in `reinit()` (the cleanest way to do things), or a `_add_p_level_on_reinit` setting in `FEAbstract` (the more MOOSE-friendly way to do things, since we could mark just that one FELagrange with it).  Hopefully the reason was just "laziness"?  We also have an optional `p_level` argument to override `elem->p_level()` in `DofMap::dof_indices()`

lindsayad committed:
subproblem -> fe_problem in Adaptivity class

lindsayad committed:
Refactor helpers in Assembly to accomodate p-refinement

roystgnr commented on test/tests/dgkernels/2d_diffusion_dg/tests:
```diff
@@ -9,6 +9,15 @@
   []
+  [p_refinement_test]
+    type = 'Exodiff'
+    input = '2d_diffusion_dg_test.i'
+    cli_args = 'Outputs/file_base=p_refinement Executioner/Adaptivity/switch_h_to_p_refinement=true'
```
That sounds like a good idea; at first glance it looks like that input file has `family=MONOMIAL` (so it should go to high order trivially), and on such a tiny mesh I wouldn't worry about conditioning issues.

roystgnr commented on test/tests/dgkernels/2d_diffusion_dg/tests:
```diff
@@ -9,6 +9,15 @@
+    input = '2d_diffusion_dg_test.i'
+    cli_args = 'Outputs/file_base=p_refinement Executioner/Adaptivity/switch_h_to_p_refinement=true'
+    exodiff = 'p_refinement.e-s003'
+    max_parallel = 1
```
This is somewhat worrisome.  Every libMesh `make check` does p-refinement examples in parallel; if we can't do the same in MOOSE too then there's probably some easy bug to fix.

lindsayad replied:
I've removed the restriction

lindsayad committed:
Add gold file for test

lindsayad committed:
Remove serial restriction on some tests

lindsayad committed:
Add 3D test

lindsayad committed:
Fix valgrind leak of helper objects

lindsayad committed:
Input file fixes

- formatting
- 'value' to 'expression' for ParsedGradFunction

lindsayad committed:
Don't add_p_level_in_reinit when computing max qps

This is consistent with previous behavior of only using first
order Lagrange

lindsayad committed:
Don't add p level when querying the number of shape functions

Once libmesh/libmesh#3597 is in, then we can use that. But until
then we can do this clumsy thing

moosebuild commented on libmesh:
```diff
@@ -1 +1 @@
-Subproject commit dc964ef4e459ba7ce080f8d6c1f4cac939cfbfed
+Subproject commit 9e95b9b6f6a7a3bb158805a782aaeebdb6b38a8d
```
**Caution!** This contains a submodule update

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1619418/) on 118f449 wanted to post the following:

A change of the following file(s) triggered this check:

  libmesh

The following file(s) are unchanged:

  conda/mpich/conda_build_config.yaml
  conda/libmesh/meta.yaml

The libmesh submodule or configuration was changed but the conda build config was not

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1619549/) on 3c779f0 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>080096</th>
      <th colspan=4>#24180 3c779f</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.58%</td>
      <td>85.59%</td>
      <td>+0.01%</td>
      <td>90.45%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>89021</td>
      <td>89135</td>
      <td>+114</td>
      <td>180</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>14997</td>
      <td>15008</td>
      <td>+11</td>
      <td>19</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24180/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24180/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/24180/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/24180/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/24180/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/24180/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/24180/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/24180/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/24180/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/24180/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/24180/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/24180/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/24180/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/24180/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/24180/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/24180/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/24180/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/24180/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/24180/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/24180/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/24180/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/24180/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/24180/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/24180/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/24180/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/24180/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/24180/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/24180/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/24180/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/24180/coverage/xfem)

</details>



This comment will be updated on new commits.

roystgnr commented on framework/src/loops/MaxQpsThread.C:
```diff
@@ -64,19 +64,41 @@ MaxQpsThread::operator()(const ConstElemRange & range)
+    // polynomial basis is higher, but our fe_type up above is first Lagrange, so it seems like the
+    // person who wrote this code didn't want to consider higher order bases even if if the user
+    // asked for higher order bases
+    fe->add_p_level_in_reinit(false);
```
I think this is wrong.

Contrary to the comment, our `fe_type` here is irrelevant, isn't it?  `attachQRuleElem` later attaches the quadrature rule to `fe`, but doesn't actually inspect `fe` to decide what quadrature rule to attach.  And then, within `libMesh::FE` itself, we don't look at the FE type when constructing quadrature points or weights - if someone has a quadratic and a cubic variable then in general they want to evaluate both on the same set of points, whatever fixed rule the user code has determined that to be ... and if the user really does want two different sets of points, they can get that through two different quadrature rules.

But!  `libMesh::Quadrature` *does* have an `unsigned int p_level = 0` argument to `init()`, which is filled with each `elem->p_level()` and which ought to be used (except for e.g. nodal quadrature rules where we ignore order) to *un*-fix the rule, to elevate the order of the rule by twice that `p_level`, via internal use of `Quadrature::get_order()`.  And we'd want to do the same here, to get an accurate count of qps.

lindsayad replied:
Maybe it's because I haven't been sleeping much, but I'm not clear on what alternative code you're suggesting

roystgnr replied:
Code?  None.  I'd just leave in a comment about how we want to retain `add_p_level_in_p_init(true)` so that our qps count here matches the number of qps that will later be allocated when assembling.

roystgnr replied:
Although now that I think about it, there's got to also be some place where we can stick a friendly assertion that `qrule.size()` doesn't exceed what `MaxQpsThread::max()` concluded, so we can have a failure of that instead of an OOB index who-knows-where.

lindsayad replied:
I'm still confused. I had to add this here such that upon recovery/restart I don't try to reinit a Lagrange FE with a p order it can't handle. You want me to remove this code or keep it as is?

roystgnr replied:
Oh, hell, we're using `LAGRANGE` in `fe` here.  Can we just switch that to `HIERARCHIC`?  We're just doing a `get_nothing()` on it, so it's probably just an overzealous assertion that's whining about the fact that it can't calculate things you never asked it to calculate.

lindsayad replied:
Ok I think what I'm going to do now that we have removed the max shape functions is just remove the FE from this routine. What do you think?

lindsayad replied:
My comment was indeed wrong, I see that now. I forgot that this is *not* the code that determines the quadrature order we use in our calculations. Indeed we are only querying the see what the maximum number of qps is based on the quadrature rules we've already set

roystgnr replied:
Remove the FE, and manually pass `elem->p_level()` to `Quadrature::init`?  Yeah, that's much smarter than my `HIERARCHIC` idea.

roystgnr commented on framework/src/loops/MaxQpsThread.C:
```diff
@@ -64,19 +64,41 @@ MaxQpsThread::operator()(const ConstElemRange & range)
     // libMesh currently allows for that
     std::unique_ptr<FEBase> side_fe(FEBase::build(dim, fe_type));
     side_fe->get_xyz();
+    side_fe->add_p_level_in_reinit(false);
```
Likewise here

roystgnr commented on framework/src/loops/MaxQpsThread.C:
```diff
@@ -64,19 +64,41 @@ MaxQpsThread::operator()(const ConstElemRange & range)
+      switch (dim)
+      {
+        case 0:
+          return static_cast<FE<0, LAGRANGE> *>(fe.get())->n_shape_functions(elem_type, FIRST);
```
And what's going on here?  While figuring out our maximum number of quadrature points (useful for pre-sizing working arrays), we also figure out the maximum number of shape functions a first-order LAGRANGE variable would have (useful for writing code that will break if ever asked to use non-trivial FEM?) too?

Could we just mark `MaxQpsThread::max_shape_funcs()` and `FEProblemBase::getMaxShapeFunctions()` with a `mooseError` and see if anything in CI needs to be fixed?  I'm not seeing any internal usage.

lindsayad replied:
I just went ahead and deleted it and CI seems happy enough

lindsayad committed:
Remove _max_shape_funcs

moosebuild commented:
Job [Python 3.10 (Apptainer)](https://civet.inl.gov/job/1633808/) on b327bb2 : invalidated by @lindsayad

lindsayad committed:
Just init quadrature rules in MaxQpsThread

roystgnr commented on framework/src/loops/MaxQpsThread.C:
```diff
@@ -47,47 +47,21 @@ MaxQpsThread::operator()(const ConstElemRange & range)
     {
-      side_fe->reinit(elem, side);
+      assem.setFaceQRule(elem, /*side=*/0);
+      const std::unique_ptr<const Elem> side(elem->build_side_ptr(0));
```
I don' think we need `side` here - `side->p_level()` is just going to be `elem->p_level()`, and `side->type()` is just going to be `elem->side_type(0)`.

But this code is correct and I wouldn't stress about a little extra heap manipulation here.

roystgnr replied:
Perhaps we should add a comment here, about assuming that side 0 always ends up with at least as many quadrature points as any other side?  That's *true*, since we number the quad side of a pyramid before the triangle sides and even the worst case triangle quadrature rules are no more expensive than a quad rule of the same order, but it's a cute benefit of libMesh numbering convention, not a priori true.

lindsayad replied:
> elem->side_type(0)

Did not know about this API!

moosebuild commented:
Job [Newer gcc](https://civet.inl.gov/job/1636798/) on 0383427 : invalidated by @lindsayad

moosebuild commented:
Job [OpenMPI](https://civet.inl.gov/job/1636800/) on 0383427 : invalidated by @lindsayad

moosebuild commented:
Job [Parallel sweep evens](https://civet.inl.gov/job/1636809/) on 0383427 : invalidated by @lindsayad

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1636785/) on 0383427 : invalidated by @lindsayad
